### ISSUE TITLE
Heap issue in Nvidia GPU - Star Swarm Benchmark

### ISSUE THREAD

**Issue by EsdrasTarsis** [libcg/grvk#8]
[memoryProperties.memoryTypeCount](https://github.com/libcg/grvk/blob/master/src/mantle/mantle_memory_man.c#L20) returns 11 heaps, but on the 10th heap the device handle changes abruptly and causes a memory access violation, it only works if I limit it to 9 heaps, below 9 heaps GR_GPU_MEMORY becomes null in grMapMemory()

### System information
- GPU: Nvidia GeForce GTX 1050 Ti
- Driver: 455.50.07
- Wine version: wine-tkg-staging-fsync-git 6.2.r5.gaa8a3d90-322 (chaotic-aur)
- GRVK version: master 

### Log files
[grvk-8-heaps.log](https://github.com/libcg/grvk/files/6044323/grvk-8-heaps.log)
[grvk-11-heaps.log](https://github.com/libcg/grvk/files/6044292/grvk-11-heaps.log)

**Comment by libcg** [comment]
Sounds like the same issue reported in https://github.com/libcg/grvk/issues/5#issuecomment-776173424

**Comment by libcg** [comment]
This is a game bug caused by the Nvidia driver exposing a bunch of memory types. Not every type supports every image format, which doesn't map well with Mantle unfortunately.

I'm not sure what's the best way to fix this yet. I could probably hide the memory types that don't expose all formats. It might work, but the performance will tank.

**Comment by sehraf** [comment]
Is it exptected that `grGetMemoryHeapCount` returns `memoryTypeCount` vs `memoryHeapCount`?

**Comment by sehraf** [comment]
What bugs me most is the fact that the device pointer goes crazy. Can this - by any chance - be a bug in Star Swam? I mean, why would the device pointer chance at all? Maybe it was not designed for so many heaps? I don't understand why Vulkan differentiates between types and heaps. My naive guess would be that Star Swam expected like 3 instead of 10.

Regarding what to do, my priority would be: Get it running first and care about performance later. Also due to the limited number of games supporting Mantle, a simple list with quirks might work just fine.

**Comment by libcg** [comment]
@sehraf Yes, the invalid pointer is caused by the excessive amount of memory heaps. After thinking about this more, I might not be able to just hide certain heaps. GRVK will need to expose less heaps to the Mantle app and route allocations to the right types behind the scenes.

**Comment by libcg** [comment]
So, it turns out Mantle only supports up to 8 heaps (GR_MAX_MEMORY_HEAPS=8), which means it's not a game bug but an API limitation. I also found out that the apps will allocate from the correct heap as indicated by `grGetObjectInfo` before calling `grAllocateMemory`. It's just a matter of hiding the right heaps and reindexing correctly.

There are 2 good candidates to eliminate out of 11 types:
- memoryTypes[3] exposes D24 types that are unused in Mantle
- memoryTypes[9] exposes cached memory
- not sure what the 3rd one would be?

**Comment by sehraf** [comment]
> memoryTypes[3] exposes D24 types that are unused in Mantle

Where do you get that information from? Looking at [VulkanCapsViewer](https://vulkan.gpuinfo.org/download.php) memory type 0 to 6 look [identical](https://vulkan.gpuinfo.org/displayreport.php?id=11032#memory) on a RTX 2060.

**Comment by libcg** [comment]
Here's a vulkaninfo dump with the relevant info: [nvidia_mem_types.txt](https://github.com/libcg/grvk/files/6360472/nvidia_mem_types.txt)
